home/views.py

```diff
@@ -14,17 +16,35 @@ def landingpage(request):
 
 def displayQuestion(request, **kwargs):
     question = Question.objects.all().get(id=kwargs['pk'])
-    sortAnsBy = request.GET["sortanswersby"] if 'sortanswersby' in request.GET else ''
+    sort_answer_by = request.GET.get("sortanswersby", "")
     context = {
         "question": question,
-        "answers": question.get_answers_feed(sortAnsBy),
+        "answers_structure": question.create_answers_structure(sort_answer_by, request.profile),
         "answersCount": question.answer_set.count(),
         "tags": question.tags.values(),
-        "title": question.get_question_title()
+        "title": question.get_question_title(),
+        "is_user_logged_in": request.user.is_authenticated
     }
     return render(request, 'home/question_detail.html', context)
 
 
+def thumbs(request):
+    if request.user.is_authenticated:
+        button = str(request.POST.get('button'))
+        (thumb_type, answer_id) = button.split(' ')
+        answer = get_object_or_404(Answer, id=answer_id)
+        profile = request.profile
+        answer.update_thumbs(profile, thumb_type)
+        question_id = answer.question.id
+
+        # HTTP_REFERER will return the url the user came from.
+        # if previous url doesn't exist (tests / incognito) then will return the second argument url
+        responseURL = request.META.get('HTTP_REFERER', reverse('question-detail', args=[question_id]))
```

Comment:
You could include the qusetion ID and answer ID in the URL to avoid having to trust `HTTP_REFERER` which can be blocked for various reasons (Because evil companies use this for spying). 

You can even make the thumb status be a part of the HTTP request and have the the Django routing mechanism do all the data validation and parsing  for you - the nice thing about Django's `url` funciton is that it can get additional arguments to build more complex URLs.

(Dong the whole thing in a GET request might expose it to CSRF attacks, not sure if that is a huge risk, but it can be mitigated by forcing it to be POST and include the CSRF token - the action URL of a form can be as complex as you want..)

Reply:
The reason I used HTTP_REFERER is not that I wanted to go back to the page displaying this question, but to get the exact URL that includes the query param the user had, for example "?sortanswersby=votes&page#answers" filtering by votes and scrolling down to answers part. I can't find a solution how to solve it otherwise

Reply:
You can reproduce the query parameters because you know all of them when you generae the page. WRT the anchor navigation (The `#` part), I'm not sure how important is that really if the answers list can be lnger then a single screen. It may be more useful to unconditionally focus the user's browser on the particular answer they "liked".

You can use `HTTP_REFERER` for extra optimization, but you must validate it like the untrustworthy piece of input that it is. Blindly redirecting users to the URL in `HTTP_REFERER` means your side can be used and an HTTP attack proxy.

Reply:
I think I'll leave this optimization alone for now

Reply:
This is a security/input validation issue not an optimization.

Reply:
I mean I'll leave the use of HTTP_REFERER  for now, will handle it later - opened issue #128  for it.

Reply:
Thanks for creating a teachable moment. This is precisely how not to defer wort to the future. Look at #128, what do you think are the chances of it ever getting prioritized? Close to zero right? Because project managers and user representative almost never care about invisible elements like this.

You can never defer security and tests work to the future, when you do:
1. Your users are at risc of relying on features that might break due to insufficient testing
2. Your users are at risc of security attacks.

So the only kind of work you can defer to the future are *user-visible features*. In this case, the right trade-off is to make the redirection a little less smart (For example, redirect to the top of the page), and defer the usage of the header to the future.

Reply:
First, thank you for the lesson. I understand now the importance of dealing with security/testing issues urgently.
I'm currently returning respond to - reverse('question-detail', args=[question_id]) which will lead to loss of query params and the user will be able to tell that his filtering is gone. so it is indeed user-visible. And I'm sure this way isn't putting the user at a security risk,